Rpc/Speaker.cs

```diff
@@ -0,0 +1,246 @@
+        var writeTask = WriteHeader(headerCts.Token);
+        var readTask = ReadHeader(headerCts.Token);
+        await TaskUtilities.CancellableWhenAll(headerCts, writeTask, readTask);
+
+        var header = RpcHeader.Parse(await readTask);
+        var expectedRole = RpcMessage<TR>.GetRole();
+        if (header.Role != expectedRole)
+            throw new ArgumentException($"Expected peer role '{expectedRole}' but got '{header.Role}'");
+
+        header.Version.Validate(ApiVersion.Current);
+    }
+
+    private async Task WriteHeader(CancellationToken ct = default)
+    {
+        var header = new RpcHeader(RpcMessage<TS>.GetRole(), ApiVersion.Current);
+        await _conn.WriteAsync(header.ToBytes(), ct);
+    }
+
+    private async Task<string> ReadHeader(CancellationToken ct = default)
+    {
+        var buf = new byte[256];
+        var have = 0;
+
+        while (true)
+        {
+            // Read into buf[have:have+1]
+            await _conn.ReadExactlyAsync(buf, have, 1, ct);
+            if (buf[have] == '\n') break;
+            have++;
+            if (have >= buf.Length)
+                throw new IOException($"Header malformed or too large: '{Encoding.UTF8.GetString(buf)}'");
+        }
+
+        return Encoding.UTF8.GetString(buf, 0, have);
+    }
+
+    private async Task ReceiveLoop(CancellationToken ct = default)
+    {
+        try
+        {
+            while (!ct.IsCancellationRequested)
+            {
+                var message = await _serdes.ReadMessage(_conn, ct);
+                if (message.RpcField.ResponseTo != 0)
+                    // Look up the TaskCompletionSource for the message ID and
+                    // complete it with the message.
+                    if (_pendingReplies.TryRemove(message.RpcField.ResponseTo, out var tcs))
+                        tcs.SetResult(message);
+
+                // TODO: we should log unknown replies
+                // Start a new task in the background to handle the message.
+                _ = Task.Run(() => Receive?.Invoke(new ReplyableRpcMessage<TS, TR>(this, message)), ct);
+            }
+        }
+        catch (OperationCanceledException)
+        {
+            // Ignore, this is expected when being disposed.
+        }
+        catch (Exception e)
+        {
+            if (!ct.IsCancellationRequested) Error?.Invoke(e);
+        }
+    }
+
+    /// <summary>
+    ///     Send a message without waiting for a reply. If a reply is received it will be handled by the callback.
+    /// </summary>
+    /// <param name="message">Message to send</param>
+    /// <param name="ct">Optional cancellation token</param>
+    public async Task SendMessage(TS message, CancellationToken ct = default)
+    {
+        using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct, _cts.Token);
+        message.RpcField = new RPC
+        {
+            MsgId = Interlocked.Add(ref _lastMessageId, 1),
```

Comment:
As discussed on our call, we need to distinguish between messages that expect a reply and those that don't in terms of something on the wire.  The original design had the RPC omitted for messages that don't expect a reply.